### ISSUE TITLE
[pickers] `AdapterMoment` shows incorrect date and day of weeks

### ISSUE THREAD

**Issue by tpmh31292** [mui/mui-x#7443]
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Steps to reproduce 🕹

Link to live example: https://codesandbox.io/s/goofy-wildflower-8r0ckf?file=/demo.js

Steps:
1. Changing the code between AdapterMoment and AdapterDayjs to see the datepicker renders differently

2. With AdapterDayjs, the picker renders date and weekday correctly
![image](https://user-images.githubusercontent.com/9924519/211241074-e79301e8-c19a-4232-ab78-58ad0bbd5d29.png)

3. With AdapterMomentjs, the picker renders correctly the first time the pickers opened, but then incorrect
![image](https://user-images.githubusercontent.com/9924519/211241214-9c56a4de-5f73-4579-ad6d-0d784995a1ec.png)

### Your environment 🌎

<details>
  <summary><code>npx @mui/envinfo</code></summary>

```
  Don't forget to mention which browser you used.
  Output from `npx @mui/envinfo` goes here.
```
</details>

**Comment by flaviendelangle** [comment]
Hi,

Thanks for your feedback
I can't reliably reproduce your problem.
Here I have the Australian locale and moment and the 1st of January is on the Sunday:

![image](https://user-images.githubusercontent.com/3309670/211308375-ee8c9c39-be9d-43ee-a0af-16ab2d36af20.png)

But sometime when I open back the picker I have your issue.
The toolbar label has also sometime the wrong locale.
I suspect it comes from how moment handle the locales.

Did you experience the same behavior outside of Codesandbox ?
I could not reproduce it in my playgrounds.

**Comment by tpmh31292** [comment]
Yes I experience it in my work project.
First time I open the picker it is correct, but if I close it and reopen it
or just switch to another locale then it shifts to incorrect.

**Comment by tpmh31292** [comment]
I think I find the problem
If I only import 1 locale, then it renders correctly (1 Jan 2023 on Sunday)
However, if I import all needed locales, then it causes the issue above
```
import "moment/locale/en-au";
// import "moment/locale/fr";
// import "moment/locale/ru";
// import "moment/locale/de";
// import "moment/locale/en-ca";
// import "moment/locale/en-gb";
// import "moment/locale/en-nz";
```

Doesn't happen with `dayjs`

**Comment by flaviendelangle** [comment]
`moment` does not have a notion of scoped locale.
From what I understand, when you import a locale, it automatically applies it.
You can update manually using `moment.locale(YOUR_LOCALE_NAME)`.

**Comment by tpmh31292** [comment]
Right, I updated the codesandbox link above

```
const selectLocale = (newLocale) => {
    setLocale(newLocale);
    moment.locale(newLocale); // adding this line
  };
```

So when user select another locale, it will pass that locale to the `LocalizationProvider` AND set the locale for moment globally.
I think it works now

**Comment by LukasTy** [comment]
I do think that there still is some sort of an issue, especially when opening date picker after having clock picker open before.
Without examining code it really seems like a totally random behaviour. 🤷 

https://user-images.githubusercontent.com/4941090/211486348-75e71017-b4a2-4094-a5e0-264641655ea9.mov

**Comment by tpmh31292** [comment]
It seems like the DatePicker didn't rerender with the new locale, so I gave it a key

`const [key, setKey] = React.useState(1);`

```
<DatePicker
   key={key}
   ....
/>
```

and update the key when changing locale
```
const selectLocale = (newLocale) => {
    setLocale(newLocale);
    setKey(key + 1);
    moment.locale(newLocale);
  };
```

It's ok now, I think.

**Comment by LukasTy** [comment]
Here is [a codesandbox](https://codesandbox.io/s/vigorous-water-d9wflt?file=/demo.js) where I think that I can no longer reproduce the issue shown above.
The main issue seems to be the fact that some locale names are unsupported by our adapter (date-io). Adding a condition to use a different (en-gb) locale when `en-au` is selected seems to solve the issue.

**Comment by tpmh31292** [comment]
@LukasTy but it is wrong though with the en-au though, because 1 January is Saturday in your codesandbox when in fact, it is Sunday

**Comment by LukasTy** [comment]
> @LukasTy but it is wrong though with the en-au though, because 1 January is Saturday in your codesandbox when in fact, it is Sunday

Oh, sorry, I did not notice the switch to a different default week day structure on Australian calendar. 🙈 
Indeed, in that case it seems we need a `key` to ensure the picker is re-rendered on locale change.
I've updated the codesandbox, then it probably boils down to only replacing `en` for `en-gb` and that solved the other inconsistency in your demo with `en` also having incorrect week days structure.
